chainerx_cc/chainerx/cuda/cuda_device/batch_norm.cc

```diff
@@ -100,154 +107,208 @@ class CudnnBNTensorDescriptor {
     cudnnTensorDescriptor_t desc_{};
 };
 
-class CudaBatchNormForwardBackward : public chainerx::GenericBatchNormForwardBackward {
+struct CudaBatchNormState : public BatchNormState {
 public:
-    explicit CudaBatchNormForwardBackward(
-            cuda_internal::CudnnHandle& cudnn_handle,
+    CudaBatchNormState(Array x_cont, Array x_mean, Array x_inv_std, Shape beta_shape, Dtype beta_dtype)
+        : x_cont_{std::move(x_cont)},
+          x_mean_{std::move(x_mean)},
+          x_inv_std_{std::move(x_inv_std)},
+          beta_shape_{std::move(beta_shape)},
+          beta_dtype_{beta_dtype} {}
+
+    const Array& x_cont() { return x_cont_; }
+    const Array& x_mean() { return x_mean_; }
+    const Array& x_inv_std() { return x_inv_std_; }
+    const Shape& beta_shape() { return beta_shape_; }
+    Dtype beta_dtype() { return beta_dtype_; }
+
+private:
+    Array x_cont_;
+    Array x_mean_;
+    Array x_inv_std_;
+    Shape beta_shape_;
+    Dtype beta_dtype_;
+};
+
+}  // namespace
+
+class CudaBatchNormOp : public BatchNormOp {
+public:
+    std::tuple<Array, std::shared_ptr<BatchNormState>> Call(
```

Comment:
Maybe better to return `std::unique_ptr` instead of `std::shared_ptr`.

Reply:
Fixed.